storage: fix replay protection for 1PC txns

andreimatei opened the pull request:
Before this patch, a "replayed" 1PC txn could succeed in committing.
This is not intended; the timestamp cache is supposed to fail an EndTxn
request if the respective txn has been finalized before. When the
request is evaluated as a 1PC, though, it was failing to perform the
required timestamp cache check. This broke possibly when we got rid of
BeginTxn.

Personally, I'm unconvinced of the importance of "replay protection" (as
opposed to idempotency where possible, and protection against races
between a commit and an abort), but we're supposed to have this
mechanism so I guess we ought to make it work. In the case of these 1PC
txns, there's also a possibility that this protection encounters a push
rather than a finalized txn, which is a good reason for this patch
(although I'm not actually sure a 1PC txn can be pushed).

Release note: None

andreimatei committed:
storage: fix replay protection for 1PC txns

Before this patch, a "replayed" 1PC txn could succeed in committing.
This is not intended; the timestamp cache is supposed to fail an EndTxn
request if the respective txn has been finalized before. When the
request is evaluated as a 1PC, though, it was failing to perform the
required timestamp cache check. This broke possibly when we got rid of
BeginTxn.

Personally, I'm unconvinced of the importance of "replay protection" (as
opposed to idempotency where possible, and protection against races
between a commit and an abort), but we're supposed to have this
mechanism so I guess we ought to make it work. In the case of these 1PC
txns, there's also a possibility that this protection encounters a push
rather than a finalized txn, which is a good reason for this patch
(although I'm not actually sure a 1PC txn can be pushed).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44566)
<!-- Reviewable:end -->

andreimatei commented:
Everything but the last commit is #44507

nvb reviewed (commented):
Thanks for addressing this! We'll want to backport this to at least v19.2.

> Reviewed 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-:-M-CT8J-1_hLQY7bgwye:b-5gv1u1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/storage/replica_proposal.go, line 712 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CRn9X2NX5NSU5CzAI:-M-CRn9X2NX5NSU5CzAJ:b-rxudbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_proposal.go#L712)):*
> ```Go
> 	// benefit.
> 	defer func() {
> 		if batch != nil {
> ```

nit: pull the condition out:
```
if batch != nil {
    defer batch.Close()
}
```
___
*[pkg/storage/replica_write.go, line 268 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSSUiBaU0nCHl19CS:-M-CSSUiBaU0nCHl19CT:b-piofks) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L268)):*
> ```Go
> 		// This transaction has been pushed. 1PC execution can't deal with that.
> 		// NOTE(Andrei): I don't think this can actually happen given that we've
> 		// passed the isOnePhaseCommit() other than through some sort of "replay".
> ```

This will certainly be able to happen in the future when we allow a transaction to acquire locks without increasing its sequence number, so we'll want to handle it.
___
*[pkg/storage/replica_write.go, line 276 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSeh_5d0K3sA33-8_:-M-CSeh_5d0K3sA33-8a:b-krfn0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L276)):*
> ```Go
> 			newTxn.WriteTimestamp.Forward(minCommitTS)
> 			return false, roachpb.NewErrorWithTxn(
> 				roachpb.NewTransactionRetryError(roachpb.RETRY_SERIALIZABLE, "Require1PC batch pushed"),
> ```

Do we need this? We already check this in `cmd_end_transaction.go`. Hmm, we also check this in `evaluateWriteBatch`. It seems like we should only need to check the flag in batch evaluation. Do you mind looking into this?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/replica_proposal.go, line 712 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CRn9X2NX5NSU5CzAI:-M-HtLHv3atCn-VCftSX:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_proposal.go#L712)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: pull the condition out:
```
if batch != nil {
    defer batch.Close()
}
```
</blockquote></details>

done
___
*[pkg/storage/replica_write.go, line 268 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSSUiBaU0nCHl19CS:-M-Hw4AiCgC9wbvexLbN:b-hhgch3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L268)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This will certainly be able to happen in the future when we allow a transaction to acquire locks without increasing its sequence number, so we'll want to handle it.
</blockquote></details>

ack, removed the note
___
*[pkg/storage/replica_write.go, line 276 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSeh_5d0K3sA33-8_:-M-HolXICC54s3NDoQFX:b-9xyqp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L276)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need this? We already check this in `cmd_end_transaction.go`. Hmm, we also check this in `evaluateWriteBatch`. It seems like we should only need to check the flag in batch evaluation. Do you mind looking into this?
</blockquote></details>

See if you like it now.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 7 of 14 files at r10, 1 of 5 files at r11, 16 of 20 files at r12, 3 of 4 files at r13, 3 of 3 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/storage/replica_write.go, line 276 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSeh_5d0K3sA33-8_:-M-LEu6J5rzFgW-n3vBe:b-wy95w4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L276)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

See if you like it now.
</blockquote></details>

We still have the same case in `evaluate1PC` though. Is that needed?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/replica_write.go, line 276 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSeh_5d0K3sA33-8_:-M-LH_Qz5zWLipxQKVnW:b3b4b8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L276)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We still have the same case in `evaluate1PC` though. Is that needed?
</blockquote></details>

I think it's needed. We've elided the `EndTxn`, so it doesn't do the `Require1PC` check. Someone's got to do it before falling back to transactional evaluation.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/storage/replica_write.go, line 276 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSeh_5d0K3sA33-8_:-M-LLx4r8BqOgHWJeI2q:b-eev79p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L276)):*
> Someone's got to do it before falling back to transactional evaluation.

Why? The transactional evaluation path will do it.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/replica_write.go, line 276 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSeh_5d0K3sA33-8_:-M-LN0tA49MuU--FikL_:b245t5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L276)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> Someone's got to do it before falling back to transactional evaluation.

Why? The transactional evaluation path will do it.
</blockquote></details>

Well yeah I've considered it, but do we really want that?
If anything, I'd prefer to remove the protection from the `EndTxn` code and keep it around here where we deal with 1PC.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/storage/replica_write.go, line 276 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSeh_5d0K3sA33-8_:-M-LPhMW9Oafz3IUZ7Y4:bf6el1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L276)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Well yeah I've considered it, but do we really want that?
If anything, I'd prefer to remove the protection from the `EndTxn` code and keep it around here where we deal with 1PC.
</blockquote></details>

But then we'd need to make sure that every path that falls back to transactional execution detects the EndTxn and checks for this condition. And not just cases where we detect the conditions for 1PC txns and actually make it to `evaluate1PC`. It all seems rather fragile.

Meanwhile, the flag is basically saying "this EndTxn should never actually be evaluated - if it is, throw an error". Adding that code in `cmd_end_transaction.go` seems like the perfect place for it.

Either way, it shouldn't exist in both places.
___
*[pkg/storage/replica_write.go, line 345 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-LQZUm2wItyh_rIzwk:-M-LQZUn5lay5oqtQlWS:b-5nv6ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/e89f0f164e133cfb673c9a257982fa2c29158d77/pkg/storage/replica_write.go#L345)):*
> ```Go
> }
> 
> func (r *Replica) evaluate1PC(
> ```

This seems like a good place to discuss what "1PC" means. I think we've noticed in the past that the topic is under-documented in the code.


<!-- Sent from Reviewable.io -->

andreimatei committed:
storage: simplify handling of Require1PC txns

Before this patch, a 1PC txn that got pushed during evaluation was
returning a retriable error. This patch gets rid of that, and leaves all
Require1PC handling to the EndTxn code, which will generate a
TxnStatusError for Require1PC txns. This simplifies the code.
Handling of the retriable errors was inconsistent anyway because
Require1PC batches that were not qualifying for 1PC evaluation because
they had been pushed before the EndTxn batch were already getting the
TransactionStatus error.

There's only one user of Require1PC which already handles
TransactionStatusErrors, so I think using minimal code for supporting it
is a good idea.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/replica_write.go, line 276 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-CSeh_5d0K3sA33-8_:-M-MHE07Dtdj6zVjjihU:b-pi3qm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0acc52b0c1d5569cf4fb80ee35fcd037ed78571c/pkg/storage/replica_write.go#L276)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

But then we'd need to make sure that every path that falls back to transactional execution detects the EndTxn and checks for this condition. And not just cases where we detect the conditions for 1PC txns and actually make it to `evaluate1PC`. It all seems rather fragile.

Meanwhile, the flag is basically saying "this EndTxn should never actually be evaluated - if it is, throw an error". Adding that code in `cmd_end_transaction.go` seems like the perfect place for it.

Either way, it shouldn't exist in both places.
</blockquote></details>

I've prepended a commit dealing with this. PTAL.
___
*[pkg/storage/replica_write.go, line 345 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-M-LQZUm2wItyh_rIzwk:-M-MHPO4DJqdhPQMfTmC:b-fz3rsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e89f0f164e133cfb673c9a257982fa2c29158d77/pkg/storage/replica_write.go#L345)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This seems like a good place to discuss what "1PC" means. I think we've noticed in the past that the topic is under-documented in the code.
</blockquote></details>

Done in the other PR.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 4 files at r17, 5 of 5 files at r18, 3 of 3 files at r19.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-:-M-QmRLU0jkY0Kws0W8X:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44566#-:-M-WUAhF262ZAtT7BL1X:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->
